cli/lock.ml

```diff
@@ -189,25 +192,72 @@ let display_verbose_diagnostics = function
   | None -> false
   | Some l -> l >= Logs.Info
 
-let interpret_solver_error ~switch_state = function
+let interpret_solver_error ~repositories solver = function
   | `Msg _ as err -> err
   | `Diagnostics d ->
-      (match Opam_solve.not_buildable_with_dune d with
+      (match Opam_solve.not_buildable_with_dune solver d with
       | [] -> ()
       | offending_packages ->
-          check_dune_universe_repo ~switch_state offending_packages);
+          check_dune_universe_repo ~repositories offending_packages);
       let verbose = display_verbose_diagnostics (Logs.level ()) in
-      Opam_solve.diagnostics_message ~verbose d
+      Opam_solve.diagnostics_message ~verbose solver d
 
-let calculate_opam ~build_only ~allow_jbuilder ~local_opam_files ~ocaml_version
-    ~target_packages =
+let pull_repository url =
+  match (OpamUrl.local_dir url, url.backend) with
+  | Some path, `rsync ->
```

Comment:
Yeah documenting it would be nice, even maybe extracting a function to unit test would make sense here.

To clarify here we want `file://...` URLs. `local_dir` will return something for `git+file://...` as well so we need to make sure it's strictly a folder with no checkout required. Opam can tell us that by telling us what it's using to fetch such a URL, in our case, rsync!